### PR TITLE
Integrate sequence-packing scheduler into training loop and wire varlen dataset into GPT pretraining

### PR DESCRIPTION
**Pull request by ilml**
## What does this PR do?

Integrates the sequence-packing scheduler into `train_step`/`evaluate` and wires the varlen dataset into the pretraining CLI, dataloader, and dataset provider (`--use-varlen-dataset`) — for both `pretrain_gpt` and `pretrain_hybrid` (per review: `HybridModel.forward` already accepts `packed_seq_params`/`padding_mask`, so the hybrid entry point gets the same batch-fetch, forward_step, and provider wiring; note hybrid THD wiring goes beyond the original #3386, which only wired GPT). Final PR of the [#3386](https://github.com/NVIDIA/Megatron-LM/pull/3386) split series tracked in the [Dynamic Context Parallelism project](https://github.com/orgs/NVIDIA/projects/297) — after this, the full THD/sequence-packing E2E from #3386 is landed. Original changes by @xiaoyao0115 in #3386.

> **Note:** Replaces #6686, which was auto-closed when its base mirror ref was deleted. All of @jaredcasper's review comments on #6686 are addressed here: no more `getattr` defensiveness in `data_samplers.py` (direct attribute access; the dead `use_vanilla_collate_fn` read is gone), the seqlen-stats accumulator block stays at its original location in `training.py` instead of moving, and `_add_varlen_dataset_args` matches the file's argument formatting. The CLI-selector checks remain in `validate_args` because `--use-varlen-dataset`/`--sft`/`--mock-data` are not core config fields (the field-owned checks moved to `ModelParallelConfig.__post_init__` via #5902).

## Series status

Remaining PRs of the [#3386](https://github.com/NVIDIA/Megatron-LM/pull/3386) split, tracked in the [Dynamic Context Parallelism project](https://github.com/orgs/NVIDIA/projects/297). **All PRs target `main`** — no stacked bases. Reviews run in parallel; merges go in order:

| Order | PR | Content |
|---|---|---|
| 1 | #5902 | config field, CLI arg, validation |
| 2 | #6734 | mock SFT + varlen datasets |
| 3 | **this PR** (#6742) | training-loop + varlen wiring (`pretrain_gpt` + `pretrain_hybrid`) |

#6734 carries a temporary copy of #5902's commit, and #6742 carries #6734's commits (review its last three commits) — the duplicates drop out automatically as parents merge and the branches are rebased.

Merged so far: #5901, #5903, #6625, #5908, #6684, #5902, #6734. This is the final PR of the series.

> **Diff note:** rebased onto main after #6734 merged — Files-changed now shows exactly this PR's three commits (training-loop integration, varlen wiring, pretrain_hybrid wiring).

## Contribution process

- [x] Draft PR per contributing guidelines
- [x] Commits signed off (DCO)

### CODE REVIEW

File: megatron/training/datasets/data_samplers.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def build_pretraining_data_loader(dataset, consumed_samples):
             data_parallel_rank=mpu.get_data_parallel_rank(),
             data_parallel_size=mpu.get_data_parallel_world_size())
     elif args.dataloader_type == 'single':
-        if args.hybrid_context_parallel:
+        if args.hybrid_context_parallel and args.sequence_packing_scheduler is None:
```
Comment (cuichenx) [inline megatron/training/datasets/data_samplers.py:55]: not in this PR but we should change everything to "dynamic" now?

**Review by cuichenx — APPROVED** [review]

**Review by zhongbozhu — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

**Comment by ilml** [comment]
/ok to test [97e73af](https://github.com/NVIDIA/Megatron-LM/pull/6742/commits/97e73af0984327eb755f472723069764dc2595ab)
